packages/cli/src/utils/readStdin.ts

```diff
@@ -4,16 +4,27 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+import { clearTimeout } from 'node:timers';
+
 export async function readStdin(): Promise<string> {
   const MAX_STDIN_SIZE = 8 * 1024 * 1024; // 8MB
   return new Promise((resolve, reject) => {
     let data = '';
     let totalSize = 0;
     process.stdin.setEncoding('utf8');
 
+    const pipedInputShouldBeAvailableInMs = 500;
+    const pipedInputTimeout = setTimeout(() => {
+      // stop reading if input is not available yet, this is needed
+      // in terminals where stdin is never TTY and nothing's piped
+      // which causes the program to get stuck expecting data from stdin
+      onEnd();
+    }, pipedInputShouldBeAvailableInMs);
+
     const onReadable = () => {
       let chunk;
       while ((chunk = process.stdin.read()) !== null) {
+        clearTimeout(pipedInputTimeout);
```

Comment:
nit: mark that the timeout has already been cleared rather than repeatedly clearing it each chunk. if nothing else that makes it clear we are only intending to do it once.